src/data-types/mailstream.c

```diff
@@ -246,12 +218,13 @@ ssize_t mailstream_read(mailstream * s, void * buf, size_t count)
   }
 
   if (left > s->buffer_max_size) {
-    read_bytes = read_through_buffer(s, cur_buf, left);
+    read_bytes = mailstream_low_read(s->low, cur_buf, left);
```

Comment:
This function (read_through_buffer) is always trying to read exactly \count bytes. If the data comes less than needed, the call will end with timeout.  Instead, we may wait for the first piece of data and return it to the caller. If the caller needs some more data, mailstream_read() will be called again.